proxy.go

```diff
@@ -499,6 +499,42 @@ func (rp *reverseProxy) refreshCacheMetrics() {
 	}
 }
 
+func (rp *reverseProxy) getUser(u **user, c **cluster, cu **clusterUser, name string) bool {
+	rp.lock.RLock()
+	defer rp.lock.RUnlock()
+
+	*u = rp.users[name]
+	if *u != nil {
+		// c and cu for toCluster and toUser must exist if applyConfig
+		// is correct.
+		// Fix applyConfig if c or cu equal to nil.
+		*c = rp.clusters[(*u).toCluster]
+		*cu = (*c).users[(*u).toUser]
+		return true
+	}
+	return false
+}
+
+func (rp *reverseProxy) getTransparentUser(u **user, c **cluster, cu **clusterUser, name string, password string) bool {
+	rp.lock.RLock()
+	defer rp.lock.RUnlock()
+
+	*u = rp.users[config.TransparentUser]
+	if *u != nil {
+		*c = rp.clusters[(*u).toCluster]
+		if *c != nil {
+			transparentCu := (*c).users[(*u).toUser]
+			if transparentCu != nil {
+				*cu = transparentCu
+				(*cu).name = name
```

Comment:
There is one thing that worries me with this transparent user concept.

What if the user has a default CH user? I know that it's a bad habit to keep it but I bet many companies are keeping it and some of them without passwords. If the default user has a pwd then CH proxy is a door to bruteforce the pwd. If the default user doesn't have pwd then it's open bar.

So you should return false if the cluster user name is empty

Reply:
Agree.